nvtabular/ops/groupby.py

```diff
@@ -0,0 +1,229 @@
+        _conv_aggs = _columns_out_from_aggs(_conv_aggs, name_sep=self.name_sep)
+
+        return list(set(self.groupby_cols) | set(_list_aggs) | set(_conv_aggs))
+
+
+def _columns_out_from_aggs(aggs, name_sep="_"):
+    # Helper function for `output_column_names`
+    _agg_cols = []
+    for k, v in aggs.items():
+        for _v in v:
+            _agg_cols.append(name_sep.join([k, _v]))
+    return _agg_cols
+
+
+def _apply_conventional_aggs(_df, groupby_cols, _conv_aggs, name_sep="_"):
+    df_cv = None  # Default return
+    if _conv_aggs:
+        _columns = list(set(groupby_cols) | set(_conv_aggs))
+        df_cv = _df[_columns].groupby(groupby_cols).agg(_conv_aggs).reset_index()
+        df_cv.columns = [
+            name_sep.join([n for n in name if n != ""]) for name in df_cv.columns.to_flat_index()
+        ]
+    return df_cv
+
+
+def _apply_list_aggs(_df, groupby_cols, _list_aggs, name_sep="_"):
+    df_la = None  # Default return
+    if _list_aggs:
+
+        # Perform initial "collect" aggregation
+        _columns = list(set(groupby_cols) | set(_list_aggs))
+        df_la = _df[_columns].groupby(groupby_cols).collect().reset_index()
+        columns = [c + f"{name_sep}list" if c in _list_aggs else c for c in df_la.columns]
+        df_la.columns = columns
+
+        # Handle "first" and "last" aggregations
+        for col, aggs in _list_aggs.items():
+            for _agg in aggs:
+                if _agg in ("first", "last"):
+                    df_la[f"{col}{name_sep}{_agg}"] = _first_or_last(
+                        df_la[f"{col}{name_sep}list"], _agg
+                    )
+            if "list" not in aggs:
+                df_la.drop(columns=[col + f"{name_sep}list"], inplace=True)
+    return df_la
+
+
+def _get_agg_dicts(groupby_cols, list_aggs, conv_aggs, columns):
+    # Get updated aggregation dicts. This should map "__all__"
+    # to specific columns, and remove elements that are not
+    # in `columns`.
+    _allowed_cols = [c for c in columns if c not in groupby_cols]
+    _list_aggs = _ensure_agg_dict(list_aggs, _allowed_cols)
+    _conv_aggs = _ensure_agg_dict(conv_aggs, _allowed_cols)
+    return _list_aggs, _conv_aggs
+
+
+def _ensure_agg_dict(_aggs, _allowed_cols):
+    # Make sure aggregation dict has legal keys
+    if "__all__" in _aggs:
+        return {col: _aggs["__all__"] for col in _allowed_cols}
+    else:
+        return {k: v for k, v in _aggs.items() if k in _allowed_cols}
+
+
+def _is_list_agg(agg):
+    # check if `agg` is a supported list aggregation
+    return agg in ("list", "first", "last")
+
+
+def _first_or_last(x, kind):
+    # Redirect to _first or _last
+    return _first(x) if kind == "first" else _last(x)
+
+
+def _first(x):
+    # Convert each element of a list column to be the first
+    # item in the list
+    offsets = x.list._column.offsets
+    elements = x.list._column.elements
+    return [elements[offsets[i]] for i in range(0, len(offsets) - 1)]
```

Comment:
I've been wondering if we can do this without iteration over each value - and instead use something like [array broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html) for this .

What do you think about something like 


```suggestion
    return elements[offsets[:-1]]
    return elements[offsets[1:].values - 1]
```

for the _first function here

and 
```python
    return elements[offsets[1:].values - 1]
```

to get the last item in _last? This would avoid iterating over the elements 1by1 in python

Reply:
Yeah - Totally makes sense.  Thanks for suggesting this.